typescript/cli/src/rebalancer/strategy/MinAmountStrategy.ts

```diff
@@ -5,57 +5,105 @@ import type { RawBalances } from '../interfaces/IStrategy.js';
+      minAmount: number;
+      target: number;
+      isRelative: true;
+    }
+  | {
+      minAmount: bigint;
+      target: bigint;
+      isRelative: false;
+    };
 
 /**
  * Strategy implementation that rebalance based on minimum amounts
  * It ensures each chain has at least the specified minimum amount
  */
 export class MinAmountStrategy extends BaseStrategy {
-  private readonly config: ChainMap<MinAmountStrategyConfig>;
+  private readonly config: ChainMap<InferredConfig> = {};
 
   constructor(config: ChainMap<MinAmountStrategyConfig>) {
     const chains = Object.keys(config);
     super(chains);
 
     for (const chain of chains) {
-      const { buffer, minAmount } = config[chain];
+      const { minAmount, target } = config[chain];
 
-      if (minAmount < 0n) {
-        throw new Error('Minimum amount cannot be negative');
+      // check range constraints
+      if (target < minAmount) {
+        throw new Error(
+          `Target must be greater than or equal to minAmount for chain ${chain}`,
+        );
       }
 
-      if (buffer < 0n || buffer > 10_000n) {
-        throw new Error('Buffer must be between 0 and 10,000 basis points');
+      if (minAmount < 0n) {
+        throw new Error(`Minimum amount cannot be negative for chain ${chain}`);
       }
-    }
 
-    this.config = config;
+      const isRelative =
+        typeof minAmount === 'number' &&
+        typeof target === 'number' &&
+        minAmount >= 0 &&
+        minAmount <= 1 &&
+        target >= 0 &&
+        target <= 1;
+
+      this.config[chain] = {
+        minAmount,
+        target,
+        isRelative,
+      } as InferredConfig;
+    }
   }
 
   /**
-   * Gets balances categorized by surplus and deficit based on minimum amounts and buffer
-   * Buffer (in basis points) is a bigint between 1n (0.01%) and 10_000n (100.00%), default 0n
+   * Gets balances categorized by surplus and deficit based on minimum amounts and targets
+   * - For absolute values: Uses exact token amounts
+   * - For relative values: Uses percentages of total balance across all chains
    */
   protected getCategorizedBalances(rawBalances: RawBalances): {
     surpluses: Delta[];
     deficits: Delta[];
   } {
+    // Get the total balance from all chains (needed for relative calculations)
+    const total = this.chains.reduce(
+      (sum, chain) => sum + rawBalances[chain],
+      0n,
+    );
+
     return this.chains.reduce(
       (acc, chain) => {
-        const minAmount = this.config[chain].minAmount;
-        const effectiveMin =
-          (minAmount * (10_000n + this.config[chain].buffer)) / 10_000n;
+        const config = this.config[chain];
         const balance = rawBalances[chain];
+        let minAmount: bigint;
+        let targetAmount: bigint;
+
+        if (!config.isRelative) {
+          minAmount = config.minAmount;
+          targetAmount = config.target;
+        } else {
+          minAmount = BigInt(Math.floor(Number(total) * config.minAmount));
```

Comment:
Could there be an issue when transforming bigints to normal numbers?

Reply:
9007199254740991 is the largest safe integer before losing precision. Total, specially with tokens with 18 decimals might have larger values.

My recommendation would be to min amount convert it to 0 - 100 in this step and then divide by 100 (or even use 0 to 1 million as denominator for more precision, similar to how it is done on smart contracts)

Reply:
for example

total * BigInt(config.minAmount * 1_000_000n) / 1_000_000n

Reply:
I went with a BigNumber approach, works. Let me know if you feel it too expensive, I can change it to the proposedapproach